core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1933,63 +1933,78 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     */
   def merge[F2[x] >: F[x], O2 >: O](
       that: Stream[F2, O2]
-  )(implicit F: Concurrent[F2]): Stream[F2, O2] = {
-    val fstream: F2[Stream[F2, O2]] = for {
-      interrupt <- F.deferred[Unit]
-      resultL <- F.deferred[Either[Throwable, Unit]]
-      resultR <- F.deferred[Either[Throwable, Unit]]
-      otherSideDone <- F.ref[Boolean](false)
-      resultChan <- Channel.unbounded[F2, Stream[F2, O2]]
-    } yield {
-
-      def watchInterrupted(str: Stream[F2, O2]): Stream[F2, O2] =
-        str.interruptWhen(interrupt.get.attempt)
-
-      // action to signal that one stream is finished, and if it is te last one
-      // then close te queue (by putting a None in it)
-      val doneAndClose: F2[Unit] = otherSideDone.getAndSet(true).flatMap {
-        // complete only if other side is done too.
-        case true  => resultChan.close.void
-        case false => F.unit
-      }
-
-      // action to interrupt the processing of both streams by completing interrupt
-      // We need to use `attempt` because `interruption` may already be completed.
-      val signalInterruption: F2[Unit] = interrupt.complete(()).void
-
-      def go(s: Stream[F2, O2], guard: Semaphore[F2]): Pull[F2, Nothing, Unit] =
-        Pull.eval(guard.acquire) >> s.pull.uncons.flatMap {
-          case Some((hd, tl)) =>
-            val send = resultChan.send(Stream.chunk(hd).onFinalize(guard.release))
-            Pull.eval(send) >> go(tl, guard)
-          case None => Pull.done
-        }
+  )(implicit F: Concurrent[F2]): Stream[F2, O2] =
+    Stream.force {
+      // `State` describes the state of an upstream stream (`this` and `that` are both upstream streams)
+      // None            : the stream is has not yet terminated
+      // Some(Left(t))   : the stream terminated with an error
+      // Some(Right(())) : the stream terminated successfully
+      type State = Option[Either[Throwable, Unit]]
+      // `bothStates` keeps track of the state of `this` and `that` stream so we can terminate the
+      // downstream when both upstreams terminate.
+      SignallingRef.of[F2, (State, State)]((None, None)).flatMap { bothStates =>
+        // `output` is used to send chunks from upstreams to downstream. It sends streams, not chunks,
+        // as each chunk has a finalizer associated with it.
+        Channel.synchronous[F2, Stream[F2, O2]].flatMap { output =>
+          // `stopDef` is used to interrupt the streams in the case that an upstream errors, or
+          // in the case that the downstream terminates.
+          Deferred[F2, Unit].map { stopDef =>
```

Comment:
As a matter of syntax, is there any concern about the use of `for` comprehensions? We have used them to reduce code inlining. 
```suggestion
      for {
        // `bothStates` keeps track of the state of `this` and `that` stream 
        // so we can terminate downstream when both upstreams terminate.      
        bothStates <- SignallingRef.of[F2, (State, State)]((None, None))
        // `output` is used to send chunks from upstreams to downstream. 
        // It sends streams, not chunks, to tie each chunk with a finalizer
        output <- Channel.synchronous[F2, Stream[F2, O2]]
        // `stopDef` is used to interrupt the upstreams if a) any of the
        // upstreams raises an error, or b) the downstream terminates.
        stopDef <- Deferred[F2, Unit]
      } yield {
```

Reply:
I've no problems with `for` comprehensions whatsoever :+1: